scipy/stats/_quantile.py

```diff
@@ -0,0 +1,268 @@
+import numpy as np
+from scipy._lib._array_api import (xp_take_along_axis, xp_default_dtype,
+                                   xp_ravel, array_namespace)
+from scipy.stats._axis_nan_policy import _broadcast_arrays, _contains_nan
+
+
+def _quantile_iv(x, p, method, axis, nan_policy, keepdims):
+    xp = array_namespace(x, p)
+    x = xp.asarray(x)
+    p = xp.asarray(p)
+
+    if not xp.isdtype(x.dtype, ('integral', 'real floating')):
+        raise ValueError("`x` must have real dtype.")
+    if xp.isdtype(x.dtype, 'integral'):
+        x = xp.astype(x, xp_default_dtype(xp))
+
+    if not xp.isdtype(p.dtype, 'real floating'):
+        raise ValueError("`p` must have real floating dtype.")
+
+    axis_none = axis is None
+    ndim = max(x.ndim, p.ndim)
+    if axis_none:
+        x = xp_ravel(x)
+        p = xp_ravel(p)
+        axis = 0
+    elif np.iterable(axis) or int(axis) != axis:
+        message = "`axis` must be an integer or None."
+        raise ValueError(message)
+    elif (axis >= ndim) or (axis < -ndim):
+        message = "`axis` is not compatible with the shapes of the inputs."
+        raise ValueError(message)
+    axis = int(axis)
+
+    methods = {'hazen', 'interpolated_inverted_cdf', 'linear',
+               'median_unbiased', 'normal_unbiased', 'weibull'}
+    if method not in methods:
+        message = f"`method` must be one of {methods}"
+        raise ValueError(message)
+
+    contains_nans = _contains_nan(x, nan_policy, xp_omit_okay=True, xp=xp)
+
+    if keepdims not in {None, True, False}:
+        message = "If specified, `keepdims` must be True or False."
+        raise ValueError(message)
+
+    dtype = xp.result_type(p, x)
+
+    if x.shape[axis] == 0:
+        shape = list(x.shape)
+        shape[axis] = 1
+        x = xp.full(shape, xp.asarray(xp.nan, dtype=dtype))
+
+    y = xp.sort(x, axis=axis)
+    y, p = _broadcast_arrays((y, p), axis=axis)
+
+    if (keepdims is False) and (p.shape[axis] != 1):
+        message = "`keepdims` may be False only if the length of `p` along `axis` is 1."
+        raise ValueError(message)
+    keepdims = (p.shape[axis] != 1) if keepdims is None else keepdims
+
+    y = xp.moveaxis(y, axis, -1)
+    p = xp.moveaxis(p, axis, -1)
+
+    if contains_nans:
```

Comment:
Probably should only do this with `nan_policy='omit'`. `propagate` will just return NaNs, so it doesn't affect the result, but conceptually it's not right, and it's slower.